### PR TITLE
ci: dockerize web application

### PR DESCRIPTION
**Pull request by KeisukeYamashita**
# Overview

I've Dockerized the web application to be portable and deployable to container runtimes (e.g. Cloud Run, Kubernetes). We aim to host the web on Cloud Run.

## What I've done

I've containerised the application and implemented the deployment workflow.

## What I haven't done

I haven't decommissioned our GCS stuff yet. 
I've implemented it separately from the existing system so that we can deploy, test, and switch over without any impact.

## How I tested

I've built in manually:

```console
$ docker build . -t reearth-visualizer-web && docker run -e AUTH0_AUDIENCE="https://api.cms.dev.reearth.io" -e AUTH0_CLIENT_ID="xxx" -e AUTH0_DOMAIN="https://reearth-dev.auth0.com" -p 8080:8080 reearth-visualizer-web
```

## Memo

- Re:Earth CMS version: https://github.com/reearth/reearth-cms/pull/1253

## Release Notes

- **New Features**
  - Introduced a new deployment process for both server and web components, enhancing modularity.
  - Added a new job for deploying the web application to AWS.
  - Implemented a new job for deploying the application to a test environment using Cloud Run.
  - Added a new job for building the web application.

- **Improvements**
  - Updated Docker build processes with new environment variables for better management.
  - Enhanced logging configuration for Nginx with a custom JSON format.
  - Improved flexibility in Vite configuration for commit hash retrieval.

- **Chores**
  - Created a `.dockerignore` file to streamline Docker builds by excluding unnecessary files.

### CODE REVIEW

File: web/docker/40-envsubst-on-reearth-config.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        if [ -z "$value" ]; then
+            eval "export $var='\"\"'"
+        elif [ "$value" != "null" ] && ! echo "$value" | grep -qE '^\{.*\}$|^\[.*\]$'; then
+            eval "export $var='\"${value}\"'"
+        fi
+    done
+}
+
+wrap_reearth_variables "$@"
+envsubst < "$_REEARTH_CONFIG_TEMPLATE_FILE" > "$_REEARTH_CONFIG_OUTPUT_FILE"
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply web/docker/40-envsubst-on-reearth-config.sh:25]: This is out of scope this PR and I'll align with https://github.com/reearth/reearth-cms/pull/1253

File: web/docker/40-envsubst-on-reearth-config.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+wrap_reearth_variables() {
+    for var in $(env | grep '^REEARTH_VISUALIZER' | cut -d= -f1); do
+        value=$(printenv "$var")
+        if [ -z "$value" ]; then
+            eval "export $var='\"\"'"
+        elif [ "$value" != "null" ] && ! echo "$value" | grep -qE '^\{.*\}$|^\[.*\]$'; then
+            eval "export $var='\"${value}\"'"
+        fi
+    done
+}
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply web/docker/40-envsubst-on-reearth-config.sh]: This is out of scope this PR and I'll align with https://github.com/reearth/reearth-cms/pull/1253

File: .github/workflows/build_web.yml [hunk 3]
```diff
@@ -96,7 +96,8 @@
           workflow: deploy-web-nightly
           token: ${{ steps.app-token.outputs.token }}
 
-  build-docker-image:
+  # TODO: Remove later after migrating to Cloud Run
+  build-docker-image-old:
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply .github/workflows/build_web.yml:100]: This is false positive, no need.

File: .github/workflows/build_web.yml [hunk 4]
```diff
@@ -163,6 +164,76 @@
...
+              TAGS+=,${IMAGE_NAME}:latest
+            fi
+          else
+            PLATFORMS=linux/amd64
+            VERSION=$SHA
+            TAGS=$IMAGE_NAME:$NAME
+          fi
+          echo "::set-output name=platforms::$PLATFORMS"
+          echo "::set-output name=version::$VERSION"
+          echo "::set-output name=tags::$TAGS"
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply .github/workflows/build_web.yml:212]: This is out of scope this PR https://github.com/reearth/reearth-cms/pull/1253.

File: .github/workflows/ci_docker_build_push.yml [hunk 5]
```diff
@@ -83,3 +85,61 @@
...
+              TAGS+=,${IMAGE_NAME}:latest
+            fi
+          else
+            PLATFORMS=linux/amd64
+            VERSION=$SHA
+            TAGS=$IMAGE_NAME:$NAME
+          fi
+          echo "::set-output name=platforms::$PLATFORMS"
+          echo "::set-output name=version::$VERSION"
+          echo "::set-output name=tags::$TAGS"
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply .github/workflows/ci_docker_build_push.yml:134]: This is out of scope this PR.

File: .github/workflows/build_web.yml [hunk 6]
```diff
@@ -163,6 +164,76 @@ jobs:
...
+  build-docker-image:
+    name: Build and push Docker image
+    runs-on: ubuntu-latest
+    needs: [release-nightly]
+    if: ${{ github.event.inputs.name != 'blank' || github.event.inputs.new_tag != 'blank' }}
+    env:
+      IMAGE_NAME: reearth/reearth-visualizer-web
+    defaults:
+      run:
+        working-directory: server
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply .github/workflows/build_web.yml]: Fixed.

File: web/docker/reearth_config.json.template [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    "documentationUrl": $REEARTH_WEB_DOCUMENTATION_URL,
+    "earlyAccessAdmins": $REEARTH_WEB_EARLY_ACCESS_ADMINS,
+    "extensionUrls": $REEARTH_WEB_EXTENSION_URLS,
+    "ip": $REEARTH_WEB_IP,
+    "marketplaceUrl": $REEARTH_WEB_MARKETPLACE_URL,
+    "passwordPolicy": $REEARTH_WEB_PASSWORD_POLICY,
+    "plugins": $REEARTH_WEB_PLUGINS,
+    "policy": $REEARTH_WEB_POLICY,
+    "published": $REEARTH_WEB_PUBLISHED,
+}
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply web/docker/reearth_config.json.template]: Fixed!

File: web/docker/40-envsubst-on-reearth-config.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
+
+set -e
+
+# rewrite index.html to change title and favicon
+_REEARTH_HTML_FILE="/usr/share/nginx/html/index.html"
+
+# generate reearth_config.json
+_REEARTH_CONFIG_TEMPLATE_FILE="/opt/reearth/reearth_config.json.template"
+_REEARTH_CONFIG_OUTPUT_FILE="/usr/share/nginx/html/reearth_config.json"
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply web/docker/40-envsubst-on-reearth-config.sh:10]: This is out of scope this PR and I'll align with https://github.com/reearth/reearth-cms/pull/1253

File: web/docker/nginx.conf.template [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  '"request_id": "$request_id",'
+  '"request_method": "$request_method",'
+  '"request_time": "$request_time",'
+  '"request_uri": "$request_uri",'
+  '"server_name": "$server_name",'
+  '"status": "$status",'
+  '"time": "$time_iso8601"'
+'}';
+
+real_ip_header ${REAL_IP_HEADER};
```
[the remark this answers is not in the document]
Comment (KeisukeYamashita) [reply web/docker/nginx.conf.template:17]: No, this is OSS so we want to make it vendor agnostic.

**Review by rot1024 — APPROVED** [review]

**Pull request merged**